### DISCUSSION TITLE
Statistics

### DISCUSSION THREAD

**Discussion by texdraft** [shreevatsa/webWEB#15] [General]
I instrumented my program to collect statistics about Pascal code. Here are the results.

TeX:

- 14 levels of expansion
- 15674 non-empty statements
- 22602 expressions
- 202 labels
- 2998 meanings installed, comprising
  - 24 record fields,
  - 2 enumerator values,
  - 21 constants,
  - 32 types,
  - 1334 variables (including parameters),
  - 370 routines, and
  - 1215 macros (661 numeric, 241 parametric, 313 normal)

Metafont:

- 13 levels of expansion
- 15256 non-empty statements
- 23240 expressions
- 182 labels
- 3059 meanings installed, comprising
  - 16 record fields,
  - 2 enumerator values,
  - 23 constants,
  - 31 types,
  - 1770 variables (including parameters),
  - 412 routines, and
  - 805 macros (456 numeric, 156 parametric, 193 normal)

`TANGLE`:

- 8 levels of expansion
- 2089 non-empty statements
- 3077 expressions
- 36 labels
- 378 meanings installed, comprising
  - 5 record fields,
  - 2 enumerator values,
  - 12 constants,
  - 12 types,
  - 164 variables (including parameters),
  - 68 routines, and
  - 115 macros (71 numeric, 16 parametric, 28 normal)

`WEAVE`:

- 8 levels of expansion
- 4522 non-empty statements
- 6377 expressions
- 38 labels
- 585 meanings installed, comprising
  - 5 record fields,
  - 2 enumerator values,
  - 13 constants,
  - 14 types,
  - 223 variables (including parameters),
  - 93 routines, and
  - 235 macros (134 numeric, 48 parametric, 53 normal)

The [P4 compiler](https://homepages.cwi.nl/~steven/pascal/pcom.p) (converted to `WEB` by prepending `@ @p` to the source):

- 0 levels of expansion
- 4071 non-empty statements
- 5209 expressions
- 8 labels
- 751 meanings installed, comprising
  - 73 record fields,
  - 112 enumerator values,
  - 33 constants,
  - 41 types,
  - 354 variables (including parameters),
  - 137 routines, and
  - 0 macros

The counts include predefined things; the two enumerator values in TeX, Metafont, `TANGLE`, and `WEAVE` are `true` and `false`.

**Comment by shreevatsa** [comment]
Cool, thanks for sharing! I had no idea that TeX and MF were several times larger than the P4 Pascal compiler, and even WEAVE was larger than it… Is it because Pascal or P4 is so well-designed that such a small compiler is possible, or is the compiler about a "normal" size and that's just how big compilers are? I don't know much about compilers and always imagined them to be super-complex things so this is really surprising to me.

**Reply by texdraft** [reply]
I think that the disparity in size is partially explained by the fact that P4 makes far better use of the language (pointers, `new`, `with`, sets) than any of Knuth's programs do. As for `WEAVE`, note this warning in section 148:

> The macros `app`, `app1`, `app2`, and `app3 `are sequences of statements that are not enclosed with
`begin` and `end`, because such delimiters would make the Pascal program much longer.

Thus its parsing mechanism is somewhat to blame. I count about 450 statements resulting from those four macros.

I wouldn't call P4 especially well-designed, but Pascal is not a very complicated language. (The major reason for the delay in finishing Phase 2 was in getting the low-level details of expansion correct; the parser was written quickly.)

And if you are interested in compilers: The files `pascal.lisp` and `phase-2.lisp` make up about two-thirds of a Pascal compiler&mdash;the only missing parts are code generation and (optionally) more complete semantic analysis (mainly in compatibility of types). Because of Common Lisp's high-level control structures and data manipulation abilities, both files are relatively short and the parsing routines in `phase-2.lisp` quite clearly show the translation of the grammar into procedures. (Together, both files contain a little under 3000 pairs of parentheses, which should very roughly correspond to the number of statements and expressions; this count is much smaller than the 9280 in P4. I'm not trying to brag, here; I just want to suggest that the program might be useful as a demonstration of a small but nontrivial compiler-like application.)
